Ramone/smart 1018 (#33)
HorizenOfficial/evm-documentation: docs/deploy.md

```

## Deploying a Contract using Hardhat
To deploy a contract, you will use a deployment script that is specific for your smart contract. 

> Hardhat script.  Go to the script/ folder and find the following deploy.ts file:

Reviewer: ```suggestion
Hardhat script.  Go to the `script/` folder and find the following deploy.ts file:
```

### Deploy contract ERC20.sol
To deploy your contract ERC20.sol, create a folder in your Hardhat project and name it, **scripts/**. Copy this script (below) and name it **ERC20.ts**.  

Before you can deploy a contract, you need to add the import statement by running the command:

```
import {ether} from hardhat;
```

Use the deployment script for the contract, ERC20.sol:  

> ```
> const deployERC20 = async () => {
> const ERC20 = await ethers.getContractFactory("ERC20"); const erc20 = await ERC20.deploy("Test", "TST");
> await erc20.deployed();
> console.log(erc20.address); 
> };
> deployERC20();
> ```

Reviewer: ```suggestion
const deployERC20 = async () => {
   const ERC20 = await ethers.getContractFactory("ERC20"); const erc20 = await ERC20.deploy("Test", "TST");
   await erc20.deployed();
   console.log(erc20.address); 
};
deployERC20();
```

Run the command to deploy the contract:

```
npx hardhat run ./scripts/ERC20.ts
```
You will see the following execution results:

```
0x82534fBC27671eEd1F99d342CAd6be213f541a87
```

> ### Deploy contract TestPayable.sol

Reviewer: Just out of curiosity, why do we give 2 example contracts in this section?

Author: The SME wanted to show payable contract, other than that, they seem to work.

Before you can deploy a contract, you need to add the import statement by running the command:

```
import {ether} from hardhat;
```

To deploy your contract **TestPayable.sol**, create a folder in your Hardhat project and name it, **scripts/**. Copy this script (below) and name it **Payable.ts**.  

```

> const deployPayable = async () => {
> const Payable = await ethers.getContractFactory("TestPayable"); const payable = await Payable.deploy();
> await payable.deployed();
> console.log(payable.address);
> await payable.pay({ value: ethers.utils.parseEther("1") });
> console.log((await ethers.provider.getBalance(payable.address)).toString()); };
> return payable.address;
> };

Reviewer: ```suggestion
const deployPayable = async () => {
   const Payable = await ethers.getContractFactory("TestPayable"); const payable = await Payable.deploy();
   await payable.deployed();
   console.log(payable.address);
   await payable.pay({ value: ethers.utils.parseEther("1") });
   console.log((await ethers.provider.getBalance(payable.address)).toString()); };
   return payable.address;
};
```

```
Run the command to deploy the contract:

```
npx hardhat run ./scripts/Payable.ts
```

You will see the following execution results:

```
0xCE3Ab1fC99a77C4648c9c471c16d4D5e4568cE9A 1000000000000000000
```

## Interact with a Contract Instance

To interact with your deployed smart contract, create a folder in your Hardhat project and name it, **tests/**. You will add test scripts that are specific to your contract instance as well as other contracts.

### Interact with TestPayable: With ABI and Without ABI
Interact with TestPayable by passing the ABI and without passing the ABI. Copy the script (below) into your **test/** folder:

```

> const contractInteraction = async () => {
> const payableAddress = await deployPayable();
> // passing the ABI
> const payable = await ethers.getContractAt("TestPayable", payableAddress); await payable.pay({ value: ethers.utils.parseEther("1") });
> // without passing the ABI
> const [sender] = await ethers.getSigners(); await sender.sendTransaction({
> to: payableAddress,
> value: ethers.utils.parseEther("1"), });
> console.log((await ethers.provider.getBalance(payableAddress)).toString()); };
> contractInteraction();

Reviewer: ```suggestion
const contractInteraction = async () => {
   const payableAddress = await deployPayable();
   // passing the ABI
   const payable = await ethers.getContractAt("TestPayable", payableAddress); await payable.pay({ value: ethers.utils.parseEther("1") });
   // without passing the ABI
   const [sender] = await ethers.getSigners(); await sender.sendTransaction({
      to: payableAddress,
      value: ethers.utils.parseEther("1"), 
   });
   console.log((await ethers.provider.getBalance(payableAddress)).toString()); };
   contractInteraction();
```

```

Run the command to interact with this contract instance:

```
npx hardhat test
```

You will see the following execution results:

```
2000000000000000000
```

### Interact with TestPayable: Read → Call 
Interact with TestPayable by using Read. Copy the script (below) into your **test/** folder:

```
const read = async () => {
const payable = await ethers.getContractAt(
"TestPayable",
await deployPayable() );
console.log(await payable.sentence()); };
read();
```

Run the command to interact with this contract instance:

```
npx hardhat test
```
You will see the following execution results:

```
This is payable
```

### Interact with TestPayable: Call Non-Payable Function
Interact with TestPayable by calling a non-payable function along with ZEN to verify that an error is thrown (as expected). Copy the script (below) into your **test/** folder:

```

> const invalidPayment = async () => {
> const payable = await ethers.getContractAt(
> "TestPayable",
> await deployPayable() );
> // @ts-ignore
> await payable.dontPay({ value: ethers.utils.parseEther("1") }); };
> invalidPayment();

Reviewer: ```suggestion
const invalidPayment = async () => {
   const payable = await ethers.getContractAt(
      "TestPayable",
      await deployPayable() 
   );
   // @ts-ignore
   await payable.dontPay({ value: ethers.utils.parseEther("1") }); };
   invalidPayment();
```

Reviewer: Not a blocker, but if we can write the code so that we don't need the `// @ts-ignore`, that would be ideal

```

Run the command to interact with this contract instance:

```
npx hardhat test
```

You will see the following execution results:

```
/home/marko/Documents/Projects/hardhat-investigation/node_modules/@ethersproject/logg er/src.ts/index.ts:269
const error: any = new Error(message); ^

> Error: non-payable method cannot override value (operation="overrides.value", value={"type":"BigNumber","hex":"0x0de0b6b3a7640000"}, code=UNSUPPORTED_OPERATION, version=contracts 5.7.0)

Reviewer: Can you double check that this the expected result? Looks like there was an error

Author: Yes, I had Chien check for correctness and he verified that it is good to go.

at Logger.makeError (/home/marko/Documents/Projects/hardhat-investigation/node_modules/@ethersproject/logg er/src.ts/index.ts:269:28)
at Logger.throwError (/home/marko/Documents/Projects/hardhat-investigation/node_modules/@ethersproject/logg er/src.ts/index.ts:281:20)
at /home/marko/Documents/Projects/hardhat-investigation/node_modules/@ethersproject/contr acts/src.ts/index.ts:268:20
at step (/home/marko/Documents/Projects/hardhat-investigation/node_modules/@ethersproject/cont racts/lib/index.js:48:23)
at Object.next (/home/marko/Documents/Projects/hardhat-investigation/node_modules/@ethersproject/cont racts/lib/index.js:29:53)
at fulfilled (/home/marko/Documents/Projects/hardhat-investigation/node_modules/@ethersproject/cont racts/lib/index.js:20:58) {
reason: 'non-payable method cannot override value',
code: 'UNSUPPORTED_OPERATION',
operation: 'overrides.value',
value: BigNumber { _hex: '0x0de0b6b3a7640000', _isBigNumber: true }
}
```

### Interact with ERC20: Using Approve or Transfer Methods
You can interact with the ECR20 contract instance by using the approve or transfer methods to check for emit events. Copy the script (below) into your **tests/** folder:

```

> const testApproveAndTransfer = async () => {
> const testERC20 = await deployTestERC20();
> const receiver = "0xF102765ba64849f77A2AC31a58e787B23480a6E0";
> const approveReceipt = await (await testERC20.approve(receiver, 1)).wait(); const approveEvent = approveReceipt.events?.filter(
> (e) => e.event === "Approval" );
> console.log(approveEvent);
> const transferReceipt = await (await testERC20.transfer(receiver, 1)).wait(); const transferEvent = transferReceipt.events?.filter(
> (e) => e.event === "Transfer" );
> console.log(transferEvent); };
> testApproveAndTransfer();

Reviewer: ```suggestion
const testApproveAndTransfer = async () => {
   const testERC20 = await deployTestERC20();
   const receiver = "0xF102765ba64849f77A2AC31a58e787B23480a6E0";
   const approveReceipt = await (await testERC20.approve(receiver, 1)).wait(); const approveEvent = approveReceipt.events?.filter(
(e) => e.event === "Approval" );
   console.log(approveEvent);
   const transferReceipt = await (await testERC20.transfer(receiver, 1)).wait(); const transferEvent = transferReceipt.events?.filter(
(e) => e.event === "Transfer" );
   console.log(transferEvent); };
   testApproveAndTransfer();
}
```

'0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
'0x00000000000000000000000010c6b6538860f3409bd6a16bbe7d6cb7bb371cbe',
'0x000000000000000000000000f102765ba64849f77a2ac31a58e787b23480a6e0' ],
data: '0x0000000000000000000000000000000000000000000000000000000000000001',
logIndex: 0,
blockHash: '0x5d68062189c3dfd49ae3d0f99b6850f1a19d96dea82a2798e4b5706b7209c2c1',
args: [
'0x10C6b6538860f3409Bd6a16BBe7d6CB7BB371cBe', '0xF102765ba64849f77A2AC31a58e787B23480a6E0', BigNumber { value: "1" },
from: '0x10C6b6538860f3409Bd6a16BBe7d6CB7BB371cBe', to: '0xF102765ba64849f77A2AC31a58e787B23480a6E0', value: BigNumber { value: "1" }
],
decode: [Function (anonymous)],
event: 'Transfer',
eventSignature: 'Transfer(address,address,uint256)', removeListener: [Function (anonymous)],
getBlock: [Function (anonymous)],
transactionIndex: 0, blockNumber: 48, transactionHa
getTransaction: [Function (anonymous)],
getTransactionReceipt: [Function (anonymous)] }
]
```

### Interact with ERC20: EOA to EOA Transaction
You can also interact with the ECR20 contract instance by making an EOA to EOA transaction. Copy the script (below) into your **tests/** folder:

```

> const eoaToEoa = async () => {
> const receiver = "0xF102765ba64849f77A2AC31a58e787B23480a6E0"; console.log(
> `Receiver balance before: ${(
> await ethers.provider.getBalance(receiver)
> ).toString()}` );
> const [sender] = await ethers.getSigners(); await sender.sendTransaction({
> to: receiver,
> value: ethers.utils.parseEther("1"), });
> console.log(
> `Receiver balance after: ${(
> await ethers.provider.getBalance(receiver) ).toString()}`
> ); };
> eoaToEoa();

Reviewer: ```suggestion
const eoaToEoa = async () => {
   const receiver = "0xF102765ba64849f77A2AC31a58e787B23480a6E0"; 
   console.log(
`Receiver balance before: ${(
await ethers.provider.getBalance(receiver)
).toString()}` );
   const [sender] = await ethers.getSigners(); await sender.sendTransaction({
      to: receiver,
      value: ethers.utils.parseEther("1"), 
   });
   console.log(
`Receiver balance after: ${(
await ethers.provider.getBalance(receiver) ).toString()}`
); };
   eoaToEoa();
}
```

```

Run the command to interact with this contract instance:

```
npx hardhat test
```

You will see the following execution results:

```
Receiver balance before: 0
Receiver balance after: 1000000000000000000
```

> ## Connect to a Dapp

Reviewer: We spell `dApp` in a different page. Let's make sure it's consistent everywhere

Author: yes, made change from Dapp to dApp

```

## Deploying a Contract using Hardhat
To deploy a contract, you will use a deployment script that is specific for your smart contract. 

> Hardhat script.  Go to the `script/` folder and find the following deploy.ts file:

Reviewer 2: The way `Hardhat script` is placed in here, I'm not too sure about. Is there supposed to be a period before this?

### Deploy contract ERC20.sol
To deploy your contract ERC20.sol, create a folder in your Hardhat project and name it, **scripts/**. Copy this script (below) and name it **ERC20.ts**.  

Before you can deploy a contract, you need to add the import statement by running the command:

```
import {ether} from hardhat;
```

Use the deployment script for the contract, ERC20.sol:  

> const deployERC20 = async () => {

Reviewer 2: should be in a code block

   const ERC20 = await ethers.getContractFactory("ERC20"); const erc20 = await ERC20.deploy("Test", "TST");
   await erc20.deployed();
   console.log(erc20.address); 
};
deployERC20();
Run the command to deploy the contract:

```
npx hardhat run ./scripts/ERC20.ts
```
You will see the following execution results:

```
0x82534fBC27671eEd1F99d342CAd6be213f541a87
```

### Deploy contract TestPayable.sol

Before you can deploy a contract, you need to add the import statement by running the command:

```
import {ether} from hardhat;
```

To deploy your contract **TestPayable.sol**, create a folder in your Hardhat project and name it, **scripts/**. Copy this script (below) and name it **Payable.ts**.  

```

> const deployPayable = async () => {
>    const Payable = await ethers.getContractFactory("TestPayable"); const payable = await Payable.deploy();
>    await payable.deployed();
>    console.log(payable.address);
>    await payable.pay({ value: ethers.utils.parseEther("1") });
>    console.log((await ethers.provider.getBalance(payable.address)).toString()); };
>    return payable.address;
> };

Reviewer: ```suggestion
const deployPayable = async () => {
   const Payable = await ethers.getContractFactory("TestPayable"); 
   const payable = await Payable.deploy();
   await payable.deployed();
   console.log(payable.address);
   await payable.pay({ value: ethers.utils.parseEther("1") });
   console.log((await ethers.provider.getBalance(payable.address)).toString()); };
   return payable.address;
};
```

```
Run the command to deploy the contract:

```
npx hardhat run ./scripts/Payable.ts
```

You will see the following execution results:

```
0xCE3Ab1fC99a77C4648c9c471c16d4D5e4568cE9A 1000000000000000000
```

## Interact with a Contract Instance

To interact with your deployed smart contract, create a folder in your Hardhat project and name it, **tests/**. You will add test scripts that are specific to your contract instance as well as other contracts.

### Interact with TestPayable: With ABI and Without ABI
Interact with TestPayable by passing the ABI and without passing the ABI. Copy the script (below) into your **test/** folder:

```

> const contractInteraction = async () => {
>    const payableAddress = await deployPayable();
>    // passing the ABI
>    const payable = await ethers.getContractAt("TestPayable", payableAddress); await payable.pay({ value: ethers.utils.parseEther("1") });
>    // without passing the ABI
>    const [sender] = await ethers.getSigners(); await sender.sendTransaction({
>       to: payableAddress,
>       value: ethers.utils.parseEther("1"), 
>    });
>    console.log((await ethers.provider.getBalance(payableAddress)).toString()); };
>    contractInteraction();

Reviewer: ```suggestion
const contractInteraction = async () => {
   const payableAddress = await deployPayable();
   // passing the ABI
   const payable = await ethers.getContractAt("TestPayable", payableAddress); 
   await payable.pay({ value: ethers.utils.parseEther("1") });
   // without passing the ABI
   const [sender] = await ethers.getSigners(); 
   await sender.sendTransaction({
      to: payableAddress,
      value: ethers.utils.parseEther("1"), 
   });
   console.log((await ethers.provider.getBalance(payableAddress)).toString()); 
};
contractInteraction();
```

```

Run the command to interact with this contract instance:

```
npx hardhat test
```

You will see the following execution results:

```
2000000000000000000
```

### Interact with TestPayable: Read → Call 

> Interact with TestPayable by using Read. Copy the script (below) into your **test/** folder:

Reviewer 2: I would say something like "by using the **read** function", or similar

```

> const read = async () => {
> const payable = await ethers.getContractAt(
> "TestPayable",
> await deployPayable() );
> console.log(await payable.sentence()); };
> read();

Reviewer: ```suggestion
const read = async () => {
   const payable = await ethers.getContractAt(
      "TestPayable",
      await deployPayable() 
   );
   console.log(await payable.sentence()); 
};
read();
```

```

Run the command to interact with this contract instance:

```
npx hardhat test
```
You will see the following execution results:

```
This is payable
```

### Interact with TestPayable: Call Non-Payable Function
Interact with TestPayable by calling a non-payable function along with ZEN to verify that an error is thrown (as expected). Copy the script (below) into your **test/** folder:

```
const invalidPayment = async () => {
   const payable = await ethers.getContractAt(
      "TestPayable",
      await deployPayable() 
   );
   // @ts-ignore

>    await payable.dontPay({ value: ethers.utils.parseEther("1") }); };

Reviewer: ```suggestion
   await payable.dontPay({ value: ethers.utils.parseEther("1") }); 
};
```

>    invalidPayment();

Reviewer: ```suggestion
invalidPayment();
```

Error: non-payable method cannot override value (operation="overrides.value", value={"type":"BigNumber","hex":"0x0de0b6b3a7640000"}, code=UNSUPPORTED_OPERATION, version=contracts 5.7.0)
at Logger.makeError (/home/marko/Documents/Projects/hardhat-investigation/node_modules/@ethersproject/logg er/src.ts/index.ts:269:28)
at Logger.throwError (/home/marko/Documents/Projects/hardhat-investigation/node_modules/@ethersproject/logg er/src.ts/index.ts:281:20)
at /home/marko/Documents/Projects/hardhat-investigation/node_modules/@ethersproject/contr acts/src.ts/index.ts:268:20
at step (/home/marko/Documents/Projects/hardhat-investigation/node_modules/@ethersproject/cont racts/lib/index.js:48:23)
at Object.next (/home/marko/Documents/Projects/hardhat-investigation/node_modules/@ethersproject/cont racts/lib/index.js:29:53)
at fulfilled (/home/marko/Documents/Projects/hardhat-investigation/node_modules/@ethersproject/cont racts/lib/index.js:20:58) {
reason: 'non-payable method cannot override value',
code: 'UNSUPPORTED_OPERATION',
operation: 'overrides.value',
value: BigNumber { _hex: '0x0de0b6b3a7640000', _isBigNumber: true }
}
```

### Interact with ERC20: Using Approve or Transfer Methods

> You can interact with the ECR20 contract instance by using the approve or transfer methods to check for emit events. Copy the script (below) into your **tests/** folder:

Reviewer: ```suggestion
You can interact with the ERC-20 contract instance by using the approve or transfer methods to check for emit events. Copy the script (below) into your **tests/** folder:
```

```

> const testApproveAndTransfer = async () => {
>    const testERC20 = await deployTestERC20();
>    const receiver = "0xF102765ba64849f77A2AC31a58e787B23480a6E0";
>    const approveReceipt = await (await testERC20.approve(receiver, 1)).wait(); const approveEvent = approveReceipt.events?.filter(
> (e) => e.event === "Approval" );
>    console.log(approveEvent);
>    const transferReceipt = await (await testERC20.transfer(receiver, 1)).wait(); const transferEvent = transferReceipt.events?.filter(
> (e) => e.event === "Transfer" );
>    console.log(transferEvent); };
>    testApproveAndTransfer();
> }

Reviewer: ```suggestion
const testApproveAndTransfer = async () => {
   const testERC20 = await deployTestERC20();
   const receiver = "0xF102765ba64849f77A2AC31a58e787B23480a6E0";
   const approveReceipt = await (await testERC20.approve(receiver, 1)).wait(); 
   const approveEvent = approveReceipt.events?.filter(
(e) => e.event === "Approval" );
   console.log(approveEvent);
   const transferReceipt = await (await testERC20.transfer(receiver, 1)).wait(); 
   const transferEvent = transferReceipt.events?.filter(
(e) => e.event === "Transfer" );
   console.log(transferEvent); 
};
testApproveAndTransfer();
```

'0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
'0x00000000000000000000000010c6b6538860f3409bd6a16bbe7d6cb7bb371cbe',
'0x000000000000000000000000f102765ba64849f77a2ac31a58e787b23480a6e0' ],
data: '0x0000000000000000000000000000000000000000000000000000000000000001',
logIndex: 0,
blockHash: '0x5d68062189c3dfd49ae3d0f99b6850f1a19d96dea82a2798e4b5706b7209c2c1',
args: [
'0x10C6b6538860f3409Bd6a16BBe7d6CB7BB371cBe', '0xF102765ba64849f77A2AC31a58e787B23480a6E0', BigNumber { value: "1" },
from: '0x10C6b6538860f3409Bd6a16BBe7d6CB7BB371cBe', to: '0xF102765ba64849f77A2AC31a58e787B23480a6E0', value: BigNumber { value: "1" }
],
decode: [Function (anonymous)],
event: 'Transfer',
eventSignature: 'Transfer(address,address,uint256)', removeListener: [Function (anonymous)],
getBlock: [Function (anonymous)],
transactionIndex: 0, blockNumber: 48, transactionHa
getTransaction: [Function (anonymous)],
getTransactionReceipt: [Function (anonymous)] }
]
```

### Interact with ERC20: EOA to EOA Transaction
You can also interact with the ECR20 contract instance by making an EOA to EOA transaction. Copy the script (below) into your **tests/** folder:

```

> const eoaToEoa = async () => {
>    const receiver = "0xF102765ba64849f77A2AC31a58e787B23480a6E0"; 
>    console.log(
> `Receiver balance before: ${(
> await ethers.provider.getBalance(receiver)
> ).toString()}` );
>    const [sender] = await ethers.getSigners(); await sender.sendTransaction({
>       to: receiver,
>       value: ethers.utils.parseEther("1"), 
>    });
>    console.log(
> `Receiver balance after: ${(
> await ethers.provider.getBalance(receiver) ).toString()}`
> ); };
>    eoaToEoa();
> }

Reviewer: ```suggestion
const eoaToEoa = async () => {
   const receiver = "0xF102765ba64849f77A2AC31a58e787B23480a6E0"; 
   console.log(`Receiver balance before: ${(await ethers.provider.getBalance(receiver)).toString()}` );
   const [sender] = await ethers.getSigners(); 
   await sender.sendTransaction({
      to: receiver,
      value: ethers.utils.parseEther("1"), 
   });
   console.log(`Receiver balance after: ${(await ethers.provider.getBalance(receiver)).toString()}`); 
};
eoaToEoa();
```

```

## Deploying a Contract using Hardhat
To deploy a contract, you will use a deployment script that is specific for your smart contract. Go to the `script/` folder and find the following deploy.ts file:

### Deploy contract ERC20.sol
To deploy your contract ERC20.sol, create a folder in your Hardhat project and name it, **scripts/**. Copy this script (below) and name it **ERC20.ts**.  

Before you can deploy a contract, you need to add the import statement by running the command:

```
import {ether} from hardhat;
```

Use the deployment script for the contract, ERC20.sol:

```
const deployERC20 = async () => {

>    const ERC20 = await ethers.getContractFactory("ERC20"); const erc20 = await ERC20.deploy("Test", "TST");

Reviewer: ```suggestion
   const ERC20 = await ethers.getContractFactory("ERC20"); 
   const erc20 = await ERC20.deploy("Test", "TST");
```
